src/clr-angular/utils/drag-and-drop/draggable-ghost.ts

```diff
@@ -0,0 +1,46 @@
+/**
+ * Copyright (c) 2016-2017 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import {animate, style, transition, trigger} from "@angular/animations";
+import {Component, ElementRef, HostBinding, Inject} from "@angular/core";
+
+import {DomAdapter} from "../../data/datagrid/render/dom-adapter";
+
+import {DragDispatcher} from "./providers/drag-dispatcher";
+import {CUSTOM_GHOST_STATE, CustomGhostState} from "./providers/draggable-tracker";
+
+@Component({
+    selector: "clr-draggable-ghost",
+    template: `
+        <ng-content></ng-content>`,
+    host: {"[class.draggable-ghost]": "true", "[class.draggable-ghost--hidden]": "true", "[attr.focusable]": "false"},
+    animations:
+        [trigger("dragEndReset",
+                 [transition(":leave", [animate(200, style({top: "{{top}}", left: "{{left}}", opacity: 0}))])])]
+})
+export class DraggableGhost {
+    @HostBinding("@dragEndReset") dragEndResetHostAnim;
+
+    draggableGhostEl: Node;
+
+    constructor(private el: ElementRef, private dragDispatcher: DragDispatcher, private domAdapter: DomAdapter,
+                @Inject(CUSTOM_GHOST_STATE) private customGhostState: CustomGhostState) {
+        this.draggableGhostEl = el.nativeElement;
+
+        let originalLeft: string;
+        let originalTop: string;
+
+        if (customGhostState.isDirectChild) {
+            originalLeft = this.dragDispatcher.dragStartPosition.x + "px";
+            originalTop = this.dragDispatcher.dragStartPosition.y + "px";
```

Comment:
Doing this in the constructor is going to create problems down the line, I believe. Especially when you have an `*ngFor` creating a bunch of draggables, or conditional ghosts.

It's very hard to review this PR because there are several unguarded accesses to deep properties of other classes like this. Most of these properties are undefined until something happens (like drag start), and they would throw undefined errors if called at the wrong time. In other words, stuff seems to be in one class or another randomly, with little thought for separation of concerns or privacy of internal state.

In  this specific case, because you seem to magically be able to access the `dragStartPosition` in the constructor of a component, I have no idea if you should move this to ngAfterViewInit, or to a subscription, or simply to guard it if the drag hasn't started yet.